exporter/exporterhelper/internal/sized_elements_channel.go

```diff
@@ -0,0 +1,101 @@
+	// the disk. Synchronization is not required because the persistent queue doesn't access it concurrently.
+	// dequeue() will decreases the initElementsCount and initTotalSize before the used field.
+	initTotalSize     uint64
+	initElementsCount uint64
+}
+
+// newSizedElementsChannel creates a sized elements channel. Each element is assigned a size by the provided sizer.
+// chanCapacity is the capacity of the underlying channel which usually should be equal to the capacity of the queue to
+// avoid blocking the producer.
+func newSizedElementsChannel[T any](capacity int) *sizedElementsChannel[T] {
+	return &sizedElementsChannel[T]{
+		used: &atomic.Uint64{},
+		cap:  uint64(capacity),
+		ch:   make(chan T, capacity),
+	}
+}
+
+// enqueue puts the element into the queue with the given sized if there is enough capacity.
+// Returns an error if the queue is full. The callback is called before the element is committed to the queue.
+// If the callback returns an error, the element is not put into the queue and the error is returned.
+func (vcq *sizedElementsChannel[T]) enqueue(el T, size uint64, callback func() error) error {
+	if vcq.initTotalSize+vcq.used.Add(size) > vcq.cap {
+		vcq.used.Add(^(size - 1))
+		return ErrQueueIsFull
+	}
+	if callback != nil {
+		if err := callback(); err != nil {
+			vcq.used.Add(^(size - 1))
+			return err
+		}
+	}
+	vcq.ch <- el
+	return nil
+}
+
+// initBulkEnqueue puts the elements into the queue with the given size. It's used by the persistent queue to
+// initialize the queue with the elements recovered from the disk.
+func (vcq *sizedElementsChannel[T]) initBulkEnqueue(els []T, totalSize uint64) {
+	vcq.initElementsCount = uint64(len(els))
+	vcq.initTotalSize = totalSize
+	if cap(vcq.ch) < len(els) {
+		vcq.ch = make(chan T, len(els))
+	}
+	for _, el := range els {
+		vcq.ch <- el
+	}
+}
+
+// dequeue removes the element from the queue and returns it.
+// The call blocks until there is an item available or the queue is stopped.
+// The function returns true when an item is consumed or false if the queue is stopped and emptied.
+// The callback is called before the element is removed from the queue. It must return the size of the element.
+func (vcq *sizedElementsChannel[T]) dequeue(callback func(T) (size uint64)) (T, bool) {
+	el, ok := <-vcq.ch
+	if !ok {
+		return el, false
+	}
+
+	size := callback(el)
+
+	if vcq.initElementsCount > 0 {
+		vcq.initElementsCount--
+		if vcq.initElementsCount == 0 || vcq.initTotalSize < size {
+			vcq.initTotalSize = 0
+		} else {
+			vcq.initTotalSize -= size
+		}
+		return el, true
+	}
+
+	vcq.used.Add(^(size - 1))
+	return el, true
+}
+
+// shutdown closes the queue channel to initiate draining of the queue.
+func (vcq *sizedElementsChannel[T]) shutdown() {
+	close(vcq.ch)
+}
+
+func (vcq *sizedElementsChannel[T]) Size() int {
+	return int(vcq.initTotalSize + vcq.used.Load())
+}
+
+func (vcq *sizedElementsChannel[T]) Capacity() int {
+	return int(vcq.cap)
```

Comment:
return `cap(ch)`?

Reply:
It can be a different value as I mentioned above